text/0000-anonymous-variants.md

```diff
@@ -0,0 +1,357 @@
+
+The proposed syntax, while systematic, is rigid and does not make normal usage particularly easy. In particular, the usage of any anonymous variant type variant always has a type placeholder followed by two colons and then a numeral, and the type inferencer expects all the anonymous variants to be fully unambiguous. The described syntax limits the expression of possible types to those that have at least one variant, and exactly one field per variant. 
+
+The anonymous variant types themselves do not have properties that may be desired in a sum or algebraic union type, such as commutativity. Because variants are identified by their position within the anonymous variant type, switching the order of the variants is a breaking change, and numbers themselves aren't particularly indicative of what they represent. 
+
+Because of coherence rules, not providing blanket implementations of user implementable traits on anonymous variant types by their stabilization will forever prevent adding such blanket trait implementations without causing breakage. 
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+What's the method behind the mad syntax? 
+------
+
+The syntax of anonymous variant types intentionally follows that of tuples. The name of the type is wrapped in parentheses, and consists of separated type names, just like in a tuple. The only syntactic difference is using separating and trailing vertical bars instead of separating and trailing commas. Vertical bars are associated with alternation in a match context, and I decided it would be fine to maintain this intuition in a type context. 
+
+The use of numbers is similar; just like a tuple's fields are numbered in ascending order from zero, and each field has the type of that ordinal type within the tuple declaration, an anonymous variant type's variants are numbered in ascending order from zero, and each variant has the type of that ordinal type within the anonymous variant type declaration. 
+
+You might think that the syntax is unpleasant, and you'd be right in that regard. This is one of tradeoffs made to make the actual proposal in this RFC as lean as possible. Your local crates vendor should have crates which help with that once this RFC gets implemented. Alternatively, you can file new RFCs which enrich the types in this RFC with ergonomic improvements. 
```

Comment:
No, the example I have demonstrates that using anonymous variant types by themselves is practical. However, things such as matching arms by type rather than by variant number, like you wanted, can be done through macros. Currently, the ecosystem has some things in this direction, but there's no canonical sum type for all of them to converge on, so it's rather fragmented.